### PR TITLE
Add pypardiso recipe

### PR DESCRIPTION
**Pull request by cmutel**
Library for sparse matrix math using the pardiso solver in the Intel MKL library.

Recipe builds and tests successfully on my machine.

Source is **not** from original source, but from a fork which implemented complex matrices. My intention is to expand on this library eventually to support newer versions of pardiso.

I confirm that I am willing to maintain this package.

Pinging @conda-forge/help-python :)

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by isuruf** [comment]
Can you use the original source and add the changes in the fork as patches?

**Comment by cmutel** [comment]
@isuruf I switched to the original source, and after some recipe debugging, everything builds correctly.

**Comment by cmutel** [comment]
Pinging @conda-forge/help-python, I think this is ready for review and possible approval.

**Comment by haasad** [comment]
> GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.

I confirm that I want to be listed as a maintainer

**Comment by cmutel** [comment]
@conda-forge/staged-recipes @isuruf Would be nice to have this reviewed and merged :)

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:9]:
```suggestion
  url: https://github.com/haasad/PyPardisoProject/archive/refs/tags/v{{ version }}.tar.gz
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:20]:
```suggestion
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:27]:
```suggestion
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:34]: Why does this need to be constrained on Windows?
```suggestion
    - mkl
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:15]: If my other commends are valid this looks like it can be `noarch: python`:

```suggestion
  noarch: python
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:30]:
```suggestion
    - numpy
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:24]:
```suggestion
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:48]:
```suggestion
    - pytest -vv tests/
    - pip check
```

File: recipes/pypardiso/meta.yaml
Comment (chrisburr) [inline recipes/pypardiso/meta.yaml:41]:
```suggestion
    - pytest
    - pip
```

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml:31]: The latest releases of MKL changed the name of the linked library, at least on Windows, so this is necessary to get the Windows version to build correctly.

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml:15]: If there a cost to build a separate package for each version? We have already lost a decent amount of time debugging specific problems with versions and OSes, my preference would be to use the well-tested conda-forge architecture to build known good packages for each version and OS combination automatically.

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml:24]: This is necessary for the builds to work properly. See https://github.com/conda-forge/staged-recipes/pull/15106/commits/1ab96ed6cec87034da962e013767caa5eb2d6a03 and failures on previous versions.

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml:21]: Sorry, the distinction between host and build is still a bit fuzzy for me. I am worried that this will somehow break things, but we can commit it and try...

**Comment by cmutel** [comment]
@chrisburr Thanks a lot for your detailed review. I think I have addressed most of your comments, except for one which I don't know the answer to (if there is one answer). I think we can't build a `noarch` version due to the Windows requirement of using an older version of Intel MKL.

**Comment by ma-sadeghi** [comment]
Hi @cmutel, is there a reason you've hardcoded MKL 2020 for Windows?

File: recipes/pypardiso/meta.yaml
Comment (isuruf) [reply recipes/pypardiso/meta.yaml:30]: I think you need `mkl-devel` and it should work

Comment (cmutel) [reply recipes/pypardiso/meta.yaml]: Yes; I updated the recipe.

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml]: Seems to build find with `noarch`

**Comment by cmutel** [comment]
@ma-sadeghi @isuruf @chrisburr I have updated the recipe to use 0.3.1, which alleviates the need for specific MKL version pinning on Windows, and allows for a `noarch` build.

**Comment by ma-sadeghi** [comment]
@cmutel Great, thanks!

**Comment by ma-sadeghi** [comment]
@conda-forge/help-python could you please take a look at this PR? Two of our libraries depend on `pypardiso` and are currently broken. Really appreciate your help :)

**Comment by ma-sadeghi** [comment]
@cmutel I think if you list PyPI as the source url (instead of GitHub), conda-forge automatically creates a PR whenever a new release is out. Once this PR is merged you can also add an automerge bot to the conda feedstock so PRs get merged automatically if build is successful and all tests pass. This way, you don't really need to do anything for maintaining the conda-forge feedstock for `pypardiso`. That's how we do it for our two libraries `openpnm` and `porespy` and I find it very convenient that I no longer need to worry about updating the conda-forge release every time a new release is out. Anyway, just a suggestion. You just need to update the source url to
```
https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
```

**Comment by ma-sadeghi** [comment]
Hi @beckermr! Can you please mark https://anaconda.org/conda-forge/pypardiso as broken? Thanks!

**Comment by beckermr** [comment]
I will fix the outputs. Please submit a PR to conda-forge/admin-requests to mark the current builds broken.

**Comment by beckermr** [comment]
PR made: https://github.com/conda-forge/feedstock-outputs/pull/18

Once this PR is merge, bump me to merge the one above and you should be good to go.

**Comment by beckermr** [comment]
OK. All of the PRs are merged. This PR can be merged when the current reviewers are happy.

**Comment by cmutel** [comment]
@ma-sadeghi Thanks Amin. I originally used the Github link because there wasn't a pypi release, but now that there is, of course we should switch.

File: recipes/pypardiso/meta.yaml
Comment (ma-sadeghi) [inline recipes/pypardiso/meta.yaml:44]: @cmutel The builds are failing because of this line, which I think you can get rid of. It basically checks which packages are installed whose dependency is not. I think since `mkl` is installed using `conda`, `pip` fails (I think for `pip check` to succeed, it needs all dependencies installed using `pip`). Anyway, you're already testing `pypardiso` in `pytest -vv tests/`, so if there's a dependency missing, it will be caught during the `import pypardiso`.

**Review by ma-sadeghi — CHANGES_REQUESTED** [review]

File: recipes/pypardiso/meta.yaml
Comment (cmutel) [reply recipes/pypardiso/meta.yaml:44]: @conda-forge/help-python Any idea why `pip check` would fail with [missing dependencies](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=339419&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=829) after switching from the Github release to the pypi release? I don't understand hwo things could be [just fine](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=338502&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=780) with everything the same except the location of the downloaded tarball...

Comment (isuruf) [reply recipes/pypardiso/meta.yaml:44]: It's because of `mkl` in https://github.com/haasad/PyPardisoProject/blob/master/setup.py#L7

File: recipes/pypardiso/meta.yaml
Comment (isuruf) [inline recipes/pypardiso/meta.yaml:47]:
```suggestion
  home: https://github.com/haasad/PyPardisoProject
```

File: recipes/pypardiso/meta.yaml
Comment (haasad) [reply recipes/pypardiso/meta.yaml:44]: Let's use the github source for now in that case, I'll tidy it up for the next version. Probably my lack of experience with PyPI that caused this.

Comment (cmutel) [reply recipes/pypardiso/meta.yaml:44]: @isuruf Yeah, but no :) `mkl` is installed into the test environment in both the [passing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=338502&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=704) and [failing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=339419&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=752) tests. `pip` is not used to install anything, it is all conda. And pip could find mkl on its path in the passing tests...

The `pip check` suggestion came from someone who knows what they are doing, so I think it is best practice in conda-forge recipes.

@haasad The pypi upload looks fine to me, I assume you did `python setup.py sdist bdist_wheel upload` or used `twine`, there isn't much you could do wrong.

Comment (isuruf) [reply recipes/pypardiso/meta.yaml:44]: There's a pip package called `mkl` which `pip check` is looking for, but we are installing the conda package. So, the best solution is to remove `mkl` in `setup.py` and keep `mkl-service` which has `mkl` as a dependency.

Comment (haasad) [reply recipes/pypardiso/meta.yaml:44]: @cmutel I released version 0.3.2, where `mkl` is removed as direct dependency in `setup.py`: https://pypi.org/project/pypardiso/

Comment (ma-sadeghi) [reply recipes/pypardiso/meta.yaml:44]: @cmutel Could you please update `meta.yaml` to version 0.3.2 (+ the sha256)? :)

**Comment by ma-sadeghi** [comment]
@beckermr Do you know how long it takes before `pypardiso` is fully deleted from Anaconda.org? 
https://anaconda.org/conda-forge/pypardiso

**Comment by ma-sadeghi** [comment]
Hi @isuruf, could you please review this PR one more time? It seems that it's in good shape. Very much appreciate it.

File: recipes/pypardiso/meta.yaml
Comment (isuruf) [inline recipes/pypardiso/meta.yaml:25]:
```suggestion
    - libblas * *mkl
```

**Pull request merged**
